src/lambda/commands/uploadLambda.ts

```diff
@@ -0,0 +1,246 @@
+                functionNode.functionName
+            ),
+            confirm: localize('AWS.generic.response.yes', 'Yes'),
+            cancel: localize('AWS.generic.response.no', 'No'),
+        },
+        window
+    )
+
+    if (!isConfirmed) {
+        getLogger().info('ImportLambda cancelled')
+        return 'Cancelled'
+    }
+
+    if (response === uploadZipItem) {
+        return await runUploadLambdaZipFile(functionNode)
+    } else if (response === zipDirItem) {
+        return await runUploadLambdaDirectory(functionNode)
+    } else {
+        return await runUploadLambdaWithSamBuild(functionNode)
+    }
+}
+
+async function runUploadLambdaDirectory(
+    functionNode: LambdaFunctionNode,
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const parentDirArr = await window.showOpenDialog({
+        canSelectFolders: true,
+        canSelectFiles: false,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+    })
+
+    if (!parentDirArr || parentDirArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    return await zipAndUploadDirectory(functionNode, parentDirArr[0].fsPath)
+}
+
+async function runUploadLambdaZipFile(
+    functionNode: LambdaFunctionNode,
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const zipFileArr = await window.showOpenDialog({
+        canSelectFolders: false,
+        canSelectFiles: true,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+        filters: {
+            'ZIP archive': ['zip'],
+        },
+    })
+
+    if (!zipFileArr || zipFileArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    const zipFile = fs.readFileSync(zipFileArr[0].fsPath)
+
+    return await uploadZipBuffer(functionNode, zipFile)
+}
+
+async function runUploadLambdaWithSamBuild(
+    functionNode: LambdaFunctionNode,
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const parentDirArr = await window.showOpenDialog({
+        canSelectFolders: true,
+        canSelectFiles: false,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+    })
+
+    if (!parentDirArr || parentDirArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    // TODO: Detect if handler is present and error out prematurely?
+    // TODO: What should we do if no manifest (package.json, requirements.txt, etc.) is present? Directory won't be `sam build`-able.
+
+    try {
+        const invoker = getSamCliContext().invoker
+        const parentDir = parentDirArr[0]
+
+        const tempDir = await makeTemporaryToolkitFolder()
+        ExtensionDisposableFiles.getInstance().addFolder(tempDir)
```

Comment:
nit/off-topic: this should be an optional behavior of `makeTemporaryToolkitFolder`, something like `makeTemporaryToolkitFolder(disposable=true)`